src/python/pants/backend/helm/target_types.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+from abc import ABCMeta
+from dataclasses import dataclass
+from textwrap import dedent
+
+from pants.core.goals.package import OutputPathField
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    BoolField,
+    Dependencies,
+    FieldSet,
+    MultipleSourcesField,
+    SingleSourceField,
+    Target,
+    TriBoolField,
+)
+from pants.util.docutil import bin_name
+
+# -----------------------------------------------------------------------------------------------
+# Generic commonly used fields
+# -----------------------------------------------------------------------------------------------
+
+
+class HelmSkipPushField(BoolField):
+    alias = "skip_push"
+    default = False
+    help = f"If set to true, do not push this helm chart to registries when running `{bin_name()} publish`."
+
+
+# -----------------------------------------------------------------------------------------------
+# `helm_chart` target
+# -----------------------------------------------------------------------------------------------
+
+
+class HelmChartMetaSourceField(SingleSourceField):
+    alias = "chart"
+    default = "Chart.yaml"
+    expected_file_extensions = (
+        ".yaml",
+        ".yml",
+    )
+    required = False
+    help = "The chart definition file"
+
+
+class HelmChartSourcesField(MultipleSourcesField):
+    default = ("values.yaml", "templates/*.yaml", "templates/*.tpl")
+    expected_file_extensions = (".yaml", ".yml", ".tpl")
+
+
+class HelmChartDependenciesField(Dependencies):
+    pass
+
+
+class HelmChartOutputPathField(OutputPathField):
+    help = dedent(
+        """\
+        The destination folder where the final packaged chart will be located.\n
+        The final package name will still follow Helm convention, this output path will only affect the destination folder where can be found.
+        """
+    )
+
+    def value_or_default(self, *, file_ending: str | None) -> str:
+        if self.value:
+            return self.value
+        return os.path.join(self.address.spec_path.replace(os.sep, "."))
+
+
+class HelmChartLintStrictField(TriBoolField):
+    alias = "lint_strict"
+    help = "If set to true, enables strict linting of this Helm chart"
+
+
+class HelmChartTarget(Target):
```

Comment:
Could this be a Target Generator that generates targets:
- helm_meta_source (for `Chart.yaml`)
- helm_template_source (for each matching file in `templates/*.yaml` `templates/*.yml` and `templates/NOTES.txt`)
- helm_helpers_source (for each matching fiel in `templates/_*.tpl`)

That way, the different goals can select a subset of templates to work with and still minimize the boilerplate in BUILD. All charts have the same layout because that's what helm requires (so stronger than just a convention).

@Eric-Arellano maybe you can provide better pointers on how to acheive this (and if it's feasible)?

Reply:
Yeah, I guess that could be a solution... I fall a bit short on how to achieve it though, as far as I know, the base class `TargetGenerator` can only generate one target type, not multiple. So any help in that front would be more than welcomed.

Reply:
Raises the question from the PR description about how to model the Helm chart as targets. For example, does Pants actually need to track the template files individually?

For some language backends, the answer is yes; for example, in the JVM (Java/Scala), each source file is tracked individually so Pants can only compile the subset that actually depend on each other. Whereas the Go backend compiles by Go package, so `go_package` tracks all Go files in a package.

It seems fine to me to start with Helm support without these additional target types. If they turn out to be necessary in the future, they can be introduced then.

(As an aside, I recall that rules can also just implement `GenerateTargetsRequest` directly without using `TargetGenerator`. See `GoModTarget` in the Go backend which does not subclass `TargetGenerator`.)

Reply:
`helm-unittest` does use individual files, not the whole chart.

Reply:
it does at the test suite level, but a test suite can render/consume more than one template from the chart in the `templates` key of the test suite YAML file. I presume the most common expectation for any developer would be that one.

As it is now, the `HelmChartRequest` will return all the chart's sources but will not include the unit tests, so we can think of using file-level targets when adding the `helm_unittest_tests` target.

Reply:
That might be good to change sooner than later. It's fairly disruptive to change something to become a target generator: it impacts the address of everything.

Reply:
I do think that generators that create large networks of targets are worth being cautious about, for a lot of the same reasons as on #13488. Metadata should be able to live near the files it impacts, and if you were going to have per-template metadata, you wouldn't want it living in parent directories.

Reply:
Hmm. I can't envision a situation that would require any per-template metadata.